koku/api/report/queries.py

```diff
@@ -239,49 +318,103 @@ def _get_time_frame_filter(self):
                 # get last 30 days
                 end = this_hour
                 start = ReportQueryHandler.n_days_ago(this_hour, 30)
+        self.start_datetime = start
+        self.end_datetime = end
+        self._create_time_interval()
+        return (self.start_datetime, self.end_datetime, self.time_interval)
+
+    def _get_filter(self):
+        """Create dictionary for filter parameters.
+
+        Returns:
+            (Dict): query filter dictionary
+
+        """
+        filter_dict = {
+            'usage_start__gte': self.start_datetime,
+            'usage_end__lte': self.end_datetime,
+        }
+        service = self.get_group_by_data('service')
+        if not ReportQueryHandler.has_wildcard(service) and service:
+            filter_dict['cost_entry_product__product_family__in'] = service
+
+        return filter_dict
+
+    def _get_group_by(self):
+        """Create list for group_by parameters."""
+        group_by = []
+        service = self.get_group_by_data('service')
+        if service:
+            group_by.append('service')
+        return group_by
+
+    def _get_annotations(self):
+        """Create dictionary for query annotations.
+
+        Returns:
+            (Dict): query annotations dictionary
 
-        if start and end and resolution == 'daily':
-            interval = ReportQueryHandler.list_days(start, end)
-        if start and end and resolution == 'monthly':
-            interval = ReportQueryHandler.list_months(start, end)
-        interval = sorted(interval)
-        return (start, end, interval)
+        """
+        annotations = {
+            'date': self.date_trunc('usage_start'),
+            'units': Concat(self.units_key, Value(''))
+        }
+        service = self.get_group_by_data('service')
+        if service:
+            annotations['service'] = Concat(
+                'cost_entry_product__product_family', Value(''))
+
+        return annotations
 
-    def _add_interval_data(self, interval, query_data):
-        """Add the interval data if not present in cost usage data.
+    @staticmethod
+    def _group_data_by_list(group_by_list, data):
+        """Group data by list.
+
+        Args:
+            group_by_list (List): list of strings to group data by
+            data    (List): list of query results
+        Returns:
+            (Dict): dictionary of grouped query results or the original data
+        """
+        if not group_by_list:
+            return data
+
+        out_data = {}
+        curr_group = group_by_list.pop()
+        for key, group in groupby(data, lambda by: by.get(curr_group)):
```

Comment:
I'm still trying to understand what the curr_group data would really look like but the [docs](https://docs.python.org/2/library/itertools.html#itertools.groupby) say: 
```Generally, the iterable needs to already be sorted on the same key function```

is ```data``` always guaranteed to be sorted on the key function?

Reply:
curr_group is the current item in the group by list `['service', 'account'].`

The data is ordered by the database; here is a the query `print(query_data.query)`
```
SELECT DATE_TRUNC('month', "reporting_awscostentrylineitem"."usage_start" AT TIME ZONE 'UTC') AS "date", CONCAT("reporting_awscostentrylineitem"."currency_code", ) AS "units", CONCAT("reporting_awscostentryproduct"."product_family", ) AS "service", SUM("reporting_awscostentrylineitem"."unblended_cost") AS "total" FROM "reporting_awscostentrylineitem" LEFT OUTER JOIN "reporting_awscostentryproduct" ON ("reporting_awscostentrylineitem"."cost_entry_product_id" = "reporting_awscostentryproduct"."id") WHERE ("reporting_awscostentrylineitem"."usage_end" <= 2018-07-01 00:00:00+00:00 AND "reporting_awscostentrylineitem"."usage_start" >= 2018-06-01 00:00:00+00:00) GROUP BY DATE_TRUNC('month', "reporting_awscostentrylineitem"."usage_start" AT TIME ZONE 'UTC'), CONCAT("reporting_awscostentrylineitem"."currency_code", ), CONCAT("reporting_awscostentryproduct"."product_family", ) ORDER BY "date" DESC, "total" DESC
```